include/dynd/string.hpp

```diff
@@ -8,11 +8,52 @@
 #include <dynd/callable.hpp>
 
 namespace dynd {
-namespace nd {
 
-  extern DYND_API struct DYND_API string_concatenation : declfunc<string_concatenation> {
-    static callable make();
-  } string_concatenation;
+  /*
+    Concatenates `nop` strings in the `s` array, storing the result in
+    `d`
+  */
+  void string_concat(size_t nop, dynd::string *d, const dynd::string *const *s);
 
-} // namespace dynd::nd
+  /*
+    Returns the number of times needle appears in haystack.
+  */
+  intptr_t string_count(const dynd::string *const haystack,
+                        const dynd::string *const needle);
+
+  /*
+    Returns byte index of the first occurrence of needle in haystack.
+    Returns -1 if not found.
+  */
+  intptr_t string_find(const dynd::string *const haystack,
+                       const dynd::string *const needle);
+
+  /*
+    In string `src`, replace all non-overlapping appearances of
+    `old_str` with `new_str`, storing the result in `dst`.
+  */
+  void string_replace(dynd::string *const dst,
+                      const dynd::string *const src,
+                      const dynd::string *const old_str,
+                      const dynd::string *const new_str);
+
```

Comment:
I think to cover all bases, we can easily support:

- A template that supports *either* a dynd::string or a std::string (or any class that has `begin`, `end`, `size` and `resize`) ... which then calls...
- A C-like API

That avoids the problem of a lot of old-school C++ APIs that generally say "first create an instance of X (which requires some sort of copy), before you can to anything..."